Stage 3 frontend auth

MarkALevy opened the pull request.

MarkALevy committed:
commit for deployment

MarkALevy committed:
update homepage in package.json

MarkALevy committed:
update package.json

MarkALevy committed:
add favicon

MarkALevy committed:
add url to README

gennady-bars commented on README.md:
```diff
@@ -1,8 +1,2 @@
-
-- [@vitejs/plugin-react](https://github.com/vitejs/vite-plugin-react/blob/main/packages/plugin-react/README.md) uses [Babel](https://babeljs.io/) for Fast Refresh
-- [@vitejs/plugin-react-swc](https://github.com/vitejs/vite-plugin-react-swc) uses [SWC](https://swc.rs/) for Fast Refresh
+The website is available at
+[www.news-explorer.serverpit.com](www.news-explorer.serverpit.com)
```
* Please, deploy the app, and show the **correct** link in `Readme`


Now it leads to `404 - page not found` 
![image](https://github.com/user-attachments/assets/3e4b1f81-fd7f-4ae6-8bce-6aa48f1bb9aa)

gennady-bars reviewed (commented):
(**Please, expand the general comment ↓**)

Dear Mark, 

unfortunately I have to reject the project without review:


* Please, deploy the app, and show the **correct** link in `Readme`
* In ` README.md` you need to add some info about your project: technologies, how to download, how to start, how to deploy the app. And you need to add the link to your **deployed app**.

MarkALevy committed:
update README.md

MarkALevy committed:
update README.md

gennady-bars reviewed (commented):
(**Please, expand the general comment ↓**)

Dear Mark, 

unfortunately I have to reject the project without review:


* When I save articles and reload the page, the `Saved articles` page doesn't show any saved items there
* When I save articles and delete them from the `Save articles` page, there are errors https://snipboard.io/VhF36z.jpg

MarkALevy committed:
fix comments #2

gennady-bars reviewed (commented):
(**Please, expand the general comment ↓**)

Dear Mark, 

unfortunately I have to reject the project without review:


* When I log in and save some articles, they are not displayed on the  `Saved articles` page https://snipboard.io/iC8pgP.jpg 
* When I reload the page, the save articles appear, but the counter is wrong https://snipboard.io/MpZRAQ.jpg 
* When I reload the page, search the same articles, then click on the `blue` save icon to **unsave** the articles, there are errors https://snipboard.io/W2hI7w.jpg

MarkALevy committed:
fix delete from search results

MarkALevy committed:
fix comments #3

MarkALevy committed:
fix search results array

MarkALevy commented:
Dear Gennady!
Thank you for your thorough review, I am definitely learning and improving thanks to your comments.
I am not getting the first two bugs that you have mentioned,
but I fixed the third bug by comparing the search results url's to the saved articles url, and assigning identical articles with the same id.

I hope that fixing this will also fix the first two comments.

Mark

MarkALevy committed:
change routing on login

MarkALevy committed:
reload on login

MarkALevy commented:
Dear Gennady!
Thank you for your thorough review, I am definitely learning and improving thanks to your comments.
I updated the useEffect for getting saved items, and added isLogedIn as a dependency.
I fixed the third bug by comparing the search results url's to the saved articles url, and assigning identical articles with the same id.

I hope that these two modifications will fix the bugs.

Mark

gennady-bars commented on src/utils/useFormWithValidation.js:
```diff
@@ -0,0 +1,42 @@
+import React, { useCallback } from "react";
+
+// hook for form control
+export function useForm() {
+  const [values, setValues] = React.useState({});
+
+  const handleChange = (event) => {
+    const target = event.target;
+    const value = target.value;
+    const name = target.name;
+    setValues({ ...values, [name]: value });
+  };
+
+  return { values, handleChange, setValues };
+}
+
+// hook for form control and form validation
+export function useFormWithValidation() {
+  const [values, setValues] = React.useState({});
```
Please, move `useForm` and `useFormWithValidation` into a `hooks` folder because theya are **hooks**

gennady-bars commented on src/components/App/App.jsx:
```diff
@@ -12,34 +12,74 @@ import Main from "../Main/Main.jsx";
+      });
   };
 
   const handleRegistration = ({ email, password, name }) => {
```
You can create a special variable for watching **loading** states during server requests:

```js
 const [isLoading, setIsLoading] = React.useState(false);

```

And then you can change it to `true` before sending requests and return it back to `false`  in `finally` to stop loading:

And you should pass `isLoading` to every popup to change the **button text**:

```js

 buttonText={isLoading? 'Saving...' : 'Save'}

```

gennady-bars commented on src/components/App/App.jsx:
```diff
@@ -12,34 +12,74 @@ import Main from "../Main/Main.jsx";
+          ...serverError,
+          regError: "A user with this email already exists",
+        });
+        console.error("Failed to register", err);
```
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js
  function handleSubmit(request) {
    // start loading
    setIsLoading(true);
    request()
      // we need to close only in `then`
      .then(handleCloseModal) 
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
      .catch(console.error)
      // and in finally we need to stop loading
      .finally(() => setIsLoading(false));
  }

```


You need to pass the **request** into the call of `handleSubmit` like that:

```js

  // here is an example
  const handleAddItem = (item) => {
    // here we create a function that returns a promise
    const makeRequest = () => {
      // `return` lets us use a promise chain `then, catch, finally`
      return addNewItem(item).then((item) => {
        setClothingItems([item, ...clothingItems]);
      });
    };
    // here we call handleSubmit passing the request
    handleSubmit(makeRequest);
  };

```

And for the 14th sprint:

```js
 // here is an example
 function handleProfileFormSubmit(inputValues) {
  // here we create a function that returns a promise 
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally`
    return api.editProfile(inputValues).then(setCurrentUser);
  }
  // here we call handleSubmit passing the request
  handleSubmit(makeRequest);
}


```

So, this way you can remove a lot of code duplicating

gennady-bars commented on src/components/PopupWithForm/PopupWithForm.jsx:
```diff
@@ -9,6 +9,8 @@ function PopupWithForm({
+  serverError,
 }) {
   return (
     <div className={`modal ${isOpen && "modal_opened"}`}>
```
If it’s interesting for you here is a way to make a special wrapper for any `popup` with cross button, `Escape` and `overlay` listeners inside. You should create component `Popup`, store it in `Popup.js`. 

```jsx
import { useEffect } from "react";

const Popup = ({ isOpen, name, onClose, children }) => {
// here is `useEffect` for the `Escape` listener
  useEffect(() => {
    // with this we prevent adding the listener if the popup is not opened
    if (!isOpen) return;
// we should define the handler inside `useEffect`, so that it wouldn’t lose the reference to be able to remove it
    const closeByEscape = (e) => {
      if (e.key === 'Escape') {
        onClose();
      }
    }

    document.addEventListener('keydown', closeByEscape)
    // don’t forget to remove the listener in the `clean-up` function
    return () => document.removeEventListener('keydown', closeByEscape)
// here we watch `isOpen` to add the listener only when it’s opened
}, [isOpen, onClose])

// here is the overlay handler
  const handleOverlay = (e) => {
    if (e.target === e.currentTarget) {
        onClose();
    }
  }

// then we add the main wrapper with class `popup` and `popup_opened`
  return (
    <div
      className={`popup ${isOpen ? "popup_opened" : ""} popup_type_${name}`}
      onClick={handleOverlay}
    >
    {/* the container for the contents */}
      <div className='popup__container'>
        {/* here will be anything you add as `children`
*/}
        {children}
        {/* add the close button */}
        <button
          className='popup__close'
          type='button'
          onClick={onClose}
        />
      </div>
    </div>
  );
};

export default Popup;


```

And now you can use `Popup` for any popup in the project:  ` ImagePopup`, ` InfoPopup` and `PopupWithForm`.

```jsx


function PopupWithForm({isOpen, name, onClose, ...props}) {
  return (
    <Popup isOpen={isOpen} name={name} onClose={onClose}>
        <h2 className='popup__title'>{props.title}</h2>


```

You can make also `Input` for inputs and `Form` for any form in the project.

gennady-bars commented on src/components/App/App.jsx:
```diff
@@ -165,99 +225,105 @@ function App() {
-                    />
-                  </ProtectedRoute>
-                }
+                <Route
+                  path="/saved-news"
+                  element={
+                    <ProtectedRoute isLoggedInLoading={isLoggedInLoading}>
```
* The `Saved articles` page must be **protected** from unauthorized visits https://snipboard.io/m2iMK1.jpg . I can visit the `/saved-news` route

Check, please, why I'm able to visit the protected page without authorization

gennady-bars reviewed (commented):
**↓Expand to see the whole summary↓**

Hello, Mark!

Great job on submitting your project! 

**Parts I really liked in your project:**

* Good job that you **close** modals only in blocks `then`
* Good job that you `catch` possible errors at the end of server requests.
* Nice `request` function
* Nice validation hook


There are few things that need to be corrected in your project. They're mostly minor issues that are easy to fix.

**Needs correcting:**

* All comments in the code and the comments below:

* The `Saved articles` page must be **protected** from unauthorized visits https://snipboard.io/m2iMK1.jpg . I can visit the `/saved-news` route
* Please, move `useForm` and `useFormWithValidation` into a `hooks` folder because theya are **hooks**


## Number of points: 73

# Summary:
### Performance Criteria

- [] <!--22.04--> The project functionality is fully implemented according to the current stage's requirements:
  - General
    - All project links and buttons are functioning.
    - Both header states function correctly. If the user is not logged in, the header should have the "Sign in" button; and if the user is logged in, there should be no "Sign in" button. The "Saved articles" link, along with a "Log out" button, should appear in its place.
    - If the user closes the tab and then returns to the site, data is taken from local storage upon mounting the `App` component.
  - "Sign up" and "Sign in" pages
    - When clicking on the "Sign up" button in the "Sign up" popup window, a request is sent to the /signup route, provided that all input fields have been filled in correctly. If the request is successful, a popup should appear, which informs the user that they are registered and offers to log them in.
    - When clicking on the "Sign in" button, provided that all input fields have been filled in correctly, a request is sent to the `/signin` route. If the request is successful, the popup is closed.
    - Until there is valid data in all the input fields on the "Sign up" and "Sign in" pages, the form submit buttons remain disabled.
    - All forms are validated on the client side. The user cannot send a request with invalid data.
  - Search for cards
    - The preloader is visible and spinning during request execution.
    - Once the search form has been submitted successfully, a block with results appears. If nothing was found, the message "Nothing found" appears.
    - 3 cards are displayed in the results block. Clicking on the "Show more" button will render the next 3 cards.
    - When an unauthorized user clicks on the icon to save an article, the registration popup window opens.
    - After the user has been authorized, the "Save" icon becomes active in the card block.
    - When clicking the "Save" icon in the card block, a request is made to the /articles route of the API.
    - When clicking on the active "Save" icon in the card block, a request to delete the card is made. After a successful request, the card is removed from the "Saved articles" page.
  - "Saved Articles" page
    - The "Saved articles" page displays the following: username, number of articles saved, and the keywords by which articles were found.
    - Card blocks in the "Saved articles" page contain the keyword by which the card was found and the trash can icon for deleting the article.
    - When clicking on the trash icon of the card in the "Saved articles" page, a request to delete the card is made. After a successful request, the card is removed from the "Saved articles" page.
- Registration and authorization:
  - [] <!--2.52--> At least one route is protected using the `ProtectedRoute` HOC component.
  - [] <!--2.52--> When trying to access the protected route, unauthorized users are redirected to / with an open authorization popup window.
  - [x] <!--2.52--> After successful authorization, the popup window is closed.
  - [x] <!--2.52--> The / route is not protected.
  - [x] <!--2.52--> If the user was logged in and closed the tab, they can return directly to any page of the application by the URL, except for the login and registration pages.
  - [x] <!--2.52--> After a successful onSignOut() handler call, the user is redirected to /.
  - [x] <!--2.52--> The `useHistory()` hook is used correctly.
  - [x] <!--2.52--> The components `<Switch />`, `<Route />`, and `<Redirect />` are used correctly.
- [x] The interaction with the JWT token works correctly:
  - [x] <!--2.52--> The JWT token is stored in localStorage.
  - [x] <!--2.52--> The JWT token is validated by a request to the server, not just local storage.
  - [x] <!--2.52--> When you log out of the account, JWT is deleted.
- [x] <!--2.52--> A global state variable has been created that stores user data.
- Components:
  - [x] <!--2.52--> Hooks are not used inside conditional statements or loops.
  - [x] <!--2.52--> Hooks are called in a component's main function.
  - [x] <!--2.52--> For class components, effects are described inside the component lifecycle methods.
  - [x] <!--2.52--> For list items, a unique id is used instead of an array index.
  - [x] <!--2.52--> Components that use profile data are subscribed to the context
  - [x] <!--2.52--> The context is embedded in the `App` component via `CurrentUserContext.Provider`.
  - [x] <!--2.52--> A state variable has been created in the root App component that stores user data. This variable is used as a value of the context provider.
- [x] <!--2.52--> Asynchronous API requests:
  - Requests can be made through the Fetch API or by using XMLHttpRequest. Third-party libraries (such as axios or jQuery) are not used.
  - API requests are contained in a separate file.
  - The chain for processing promises ends with a `catch()` block.
  - The first `then()` handler returns `res.json`.
- [x] <!--2.52--> Naming:
  - camelCase is used for function and variable names.
  - Only nouns are used as variable names.
  - Variable names clearly describe what is stored in them. If the project has several variables with similar data, then those variables have unique but descriptive names.
  - Descriptive names are used for functions, which reflect what they do.
  - Function names start with a verb.
  - JS classes and functional components are named using nouns and start with a capital letter.
  - Names must not include inappropriate or unclear abbreviations.
  - Сustom hook names start with `use`.
- [x] <!--2.52--> No third-party JavaScript libraries are used unless absolutely necessary. If third-party libraries are connected, then they are used appropriately.

### Best Practices

- [x] <!--2.5--> The initial state of state variables contain the correct data type.
- [x] <!--2.5--> The API request for information about the user and for the array of cards is made once during mounting.
- [x] <!--2.5--> API requests are described inside the App component.
- [x] <!--2.5--> There is no memory leak when hanging handlers. All handlers added with `addEventListener` are removed when the component is unmounted.
- [x] <!--2.5--> API error handling:
  - [x] <!--2.5--> The user receives a message in the case of an error.
- [x] <!--2.5--> Non-variable values (hard-coded constants) are named in all capital letters and stored in a separate configuration file.

### Recommendations

- [x] <!--1.66--> Components from the react-router library are used for internal links in the application. 
- [x] <!--1.66--> Semantically correct blocks are used for components. No `<div />` or other unnecessary HTML tags are used for components that consist of single-level blocks.
- [x] <!--1.68--> The code is clean and easy to understand:
  - The code is readable and clearly structured. Some parts of the code are explained with comments if needed.
  - There is no extra code: for example, when a variable is declared but not used, or there is some kind of redundant logic.
  - The code is formatted in the same way, and the indentation hierarchy is respected.

## Number of points: 73

## Additional comments:
-

MarkALevy committed:
fix protected route

MarkALevy committed:
change succesful registration

gennady-bars reviewed (approved):
You’ve done a great job!

Your project has been accepted. 

Good luck!



## Number of points:  97

# Summary:
### Performance Criteria

- [x] <!--22.04--> The project functionality is fully implemented according to the current stage's requirements:
  - General
    - All project links and buttons are functioning.
    - Both header states function correctly. If the user is not logged in, the header should have the "Sign in" button; and if the user is logged in, there should be no "Sign in" button. The "Saved articles" link, along with a "Log out" button, should appear in its place.
    - If the user closes the tab and then returns to the site, data is taken from local storage upon mounting the `App` component.
  - "Sign up" and "Sign in" pages
    - When clicking on the "Sign up" button in the "Sign up" popup window, a request is sent to the /signup route, provided that all input fields have been filled in correctly. If the request is successful, a popup should appear, which informs the user that they are registered and offers to log them in.
    - When clicking on the "Sign in" button, provided that all input fields have been filled in correctly, a request is sent to the `/signin` route. If the request is successful, the popup is closed.
    - Until there is valid data in all the input fields on the "Sign up" and "Sign in" pages, the form submit buttons remain disabled.
    - All forms are validated on the client side. The user cannot send a request with invalid data.
  - Search for cards
    - The preloader is visible and spinning during request execution.
    - Once the search form has been submitted successfully, a block with results appears. If nothing was found, the message "Nothing found" appears.
    - 3 cards are displayed in the results block. Clicking on the "Show more" button will render the next 3 cards.
    - When an unauthorized user clicks on the icon to save an article, the registration popup window opens.
    - After the user has been authorized, the "Save" icon becomes active in the card block.
    - When clicking the "Save" icon in the card block, a request is made to the /articles route of the API.
    - When clicking on the active "Save" icon in the card block, a request to delete the card is made. After a successful request, the card is removed from the "Saved articles" page.
  - "Saved Articles" page
    - The "Saved articles" page displays the following: username, number of articles saved, and the keywords by which articles were found.
    - Card blocks in the "Saved articles" page contain the keyword by which the card was found and the trash can icon for deleting the article.
    - When clicking on the trash icon of the card in the "Saved articles" page, a request to delete the card is made. After a successful request, the card is removed from the "Saved articles" page.
- Registration and authorization:
  - [x] <!--2.52--> At least one route is protected using the `ProtectedRoute` HOC component.
  - [] <!--2.52--> When trying to access the protected route, unauthorized users are redirected to / with an open authorization popup window.
  - [x] <!--2.52--> After successful authorization, the popup window is closed.
  - [x] <!--2.52--> The / route is not protected.
  - [x] <!--2.52--> If the user was logged in and closed the tab, they can return directly to any page of the application by the URL, except for the login and registration pages.
  - [x] <!--2.52--> After a successful onSignOut() handler call, the user is redirected to /.
  - [x] <!--2.52--> The `useHistory()` hook is used correctly.
  - [x] <!--2.52--> The components `<Switch />`, `<Route />`, and `<Redirect />` are used correctly.
- [x] The interaction with the JWT token works correctly:
  - [x] <!--2.52--> The JWT token is stored in localStorage.
  - [x] <!--2.52--> The JWT token is validated by a request to the server, not just local storage.
  - [x] <!--2.52--> When you log out of the account, JWT is deleted.
- [x] <!--2.52--> A global state variable has been created that stores user data.
- Components:
  - [x] <!--2.52--> Hooks are not used inside conditional statements or loops.
  - [x] <!--2.52--> Hooks are called in a component's main function.
  - [x] <!--2.52--> For class components, effects are described inside the component lifecycle methods.
  - [x] <!--2.52--> For list items, a unique id is used instead of an array index.
  - [x] <!--2.52--> Components that use profile data are subscribed to the context
  - [x] <!--2.52--> The context is embedded in the `App` component via `CurrentUserContext.Provider`.
  - [x] <!--2.52--> A state variable has been created in the root App component that stores user data. This variable is used as a value of the context provider.
- [x] <!--2.52--> Asynchronous API requests:
  - Requests can be made through the Fetch API or by using XMLHttpRequest. Third-party libraries (such as axios or jQuery) are not used.
  - API requests are contained in a separate file.
  - The chain for processing promises ends with a `catch()` block.
  - The first `then()` handler returns `res.json`.
- [x] <!--2.52--> Naming:
  - camelCase is used for function and variable names.
  - Only nouns are used as variable names.
  - Variable names clearly describe what is stored in them. If the project has several variables with similar data, then those variables have unique but descriptive names.
  - Descriptive names are used for functions, which reflect what they do.
  - Function names start with a verb.
  - JS classes and functional components are named using nouns and start with a capital letter.
  - Names must not include inappropriate or unclear abbreviations.
  - Сustom hook names start with `use`.
- [x] <!--2.52--> No third-party JavaScript libraries are used unless absolutely necessary. If third-party libraries are connected, then they are used appropriately.

### Best Practices

- [x] <!--2.5--> The initial state of state variables contain the correct data type.
- [x] <!--2.5--> The API request for information about the user and for the array of cards is made once during mounting.
- [x] <!--2.5--> API requests are described inside the App component.
- [x] <!--2.5--> There is no memory leak when hanging handlers. All handlers added with `addEventListener` are removed when the component is unmounted.
- [x] <!--2.5--> API error handling:
  - [x] <!--2.5--> The user receives a message in the case of an error.
- [x] <!--2.5--> Non-variable values (hard-coded constants) are named in all capital letters and stored in a separate configuration file.

### Recommendations

- [x] <!--1.66--> Components from the react-router library are used for internal links in the application. 
- [x] <!--1.66--> Semantically correct blocks are used for components. No `<div />` or other unnecessary HTML tags are used for components that consist of single-level blocks.
- [x] <!--1.68--> The code is clean and easy to understand:
  - The code is readable and clearly structured. Some parts of the code are explained with comments if needed.
  - There is no extra code: for example, when a variable is declared but not used, or there is some kind of redundant logic.
  - The code is formatted in the same way, and the indentation hierarchy is respected.

## Number of points:  97

## Additional comments:
-
